File: src/database/connection.py
import os
from neo4j import GraphDatabase

# Singleton Driver
_DRIVER = None

def init_driver():
    global _DRIVER
    uri = os.getenv("NEO4J_URI", "bolt://localhost:7687")
    user = os.getenv("NEO4J_USER", "neo4j")
    password = os.getenv("NEO4J_PASSWORD", "password")
    
    try:
        _DRIVER = GraphDatabase.driver(uri, auth=(user, password))
        _DRIVER.verify_connectivity()
        print("Connected to Neo4j (Singleton).")
    except Exception as e:
        print(f"Failed to connect to Neo4j: {e}")
        _DRIVER = None

def get_driver():
    return _DRIVER

def close_driver():
    if _DRIVER:
        _DRIVER.close()

File: src/database/connection.js
import neo4j from 'neo4j-driver';
import { config } from '../config/env.js';
import logger from '../utils/logger.js';

let driver;

export const initDriver = async () => {
    try {
        logger.info(`Attempting to connect to Neo4j at ${config.NEO4J_URI}...`);
        driver = neo4j.driver(
            config.NEO4J_URI,
            neo4j.auth.basic(config.NEO4J_USER, config.NEO4J_PASSWORD)
        );
        await driver.verifyConnectivity();
        logger.info(`Successfully connected to Neo4j at ${config.NEO4J_URI}`);
    } catch (error) {
        logger.error(`Failed to connect to Neo4j: ${error.message}`);
        // Do not exit process, just log error. This allows retries if needed, or at least doesn't crash immediate.
        // But for this app, no DB means no work. 
        process.exit(1);
    }
};

export const getDriver = () => {
    if (!driver) {
        throw new Error('Neo4j Driver not initialized. Call initDriver() first.');
    }
    return driver;
};

export const closeDriver = async () => {
    if (driver) {
        await driver.close();
        logger.info('Neo4j Driver closed.');
    }
};
